### PR TITLE
Add bindings when computing targets

### LINKED ISSUE #1111: AntiAliasing creating unknown identifiers
Unfortunately, there is a follow up to #1097. Stainless' AntiAliasing seems to make up new identifiers with the same name as existing ones but a new id. The following (minimised) example fails with the latest version of Stainless 06b76818594c89c2d07e1d9b26aefb4f90e3c8dc with the error: `Variable key is not defined in context` (entire message below).

```scala
import stainless.annotation._

object Mutable {
  sealed abstract class Option[@mutable T]
  case class None[@mutable T]()               extends Option[T]
  case class Some[@mutable T](_0: MutCell[T]) extends Option[T]

  sealed case class Tuple2[@mutable T0, @mutable T1](_0: MutCell[T0], _1: MutCell[T1])

  case class MutCell[@mutable T](var value: T)

  def get_mut[@mutable V](self: MutCell[Tuple2[String, V]], key: String): Option[V] =
    if (self.value._0.value == key) Some[V](self.value._1)
    else None[V]()

  @pure
  def main: Unit = {
    val target = MutCell[Tuple2[String, Int]](
      Tuple2[String, Int](MutCell[String]("bar"), MutCell[Int](0))
    )
    get_mut[Int](target, "foo") match {
      case Some(v1) =>
        v1.value = 123 // if this line is commented, all passes
      case _ =>
        ()
    }
  }
}
```

<details>
<summary>Entire `--debug=stack` output</summary>

```
[info] [Warning ] Parallelism is disabled.
[info] [  Info  ] Starting verification...
[info] [  Info  ] Verified: 0 / 0
[info] [  Info  ] Verified: 0 / 0
[info] [  Info  ] Verified: 0 / 0 / compile 0s
[info] [  Info  ] Verified: 0 / 0
[info] [  Info  ] Verified: 0 / 0
[info] [ Fatal  ] ../../Mutable.scala:13:32: Variable key is not defined in context:
[info] [ Fatal  ] Kind: type-checking
[info] [ Fatal  ] Check SAT: false
[info] [ Fatal  ] Emit VCs: true
[info] [ Fatal  ] Functions: get_mut
[info] [ Fatal  ] ADTs: Tuple2, MutCell, Option
[info] [ Fatal  ] Type Variables: 
[info] [ Fatal  ] Term Variables:
[info] [ Fatal  ]   target: MutCell[Tuple2[String, Int]]
[info] [ Fatal  ]   target == MutCell[Tuple2[String, Int]](Tuple2[String, Int](MutCell[String]("bar"), MutCell[Int](0)))
[info] [ Fatal  ]   get_mut[Int](target, "foo").isInstanceOf[Some]
[info] [ Fatal  ] 
[info]                if (self.value._0.value == key) Some[V](self.value._1)
[info]                                           ^^^
[info] [ Error  ] Stainless terminated with an error.
[info] [ Error  ] Debug output is available in the file `stainless-stack-trace.txt`. If the crash is caused by Stainless, you may report your issue on https://github.com/epfl-lara/stainless/issues
[info] [ Debug  ] inox.package$FatalError: Variable key is not defined in context:
[info] [ Debug  ] Kind: type-checking
[info] [ Debug  ] Check SAT: false
[info] [ Debug  ] Emit VCs: true
[info] [ Debug  ] Functions: get_mut
[info] [ Debug  ] ADTs: Tuple2, MutCell, Option
[info] [ Debug  ] Type Variables: 
[info] [ Debug  ] Term Variables:
[info] [ Debug  ]   target: MutCell[Tuple2[String, Int]]
[info] [ Debug  ]   target == MutCell[Tuple2[String, Int]](Tuple2[String, Int](MutCell[String]("bar"), MutCell[Int](0)))
[info] [ Debug  ]   get_mut[Int](target, "foo").isInstanceOf[Some]
[info] [ Debug  ] 
[info] [ Debug  ] 	at inox.Reporter.onFatal(Reporter.scala:43)
[info] [ Debug  ] 	at inox.Reporter.fatalError(Reporter.scala:54)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.inferType(TypeChecker.scala:694)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.inferType$(TypeChecker.scala:455)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.inferType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.inferType(TypeChecker.scala:699)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.inferType$(TypeChecker.scala:455)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.inferType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1129)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1110)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1093)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1111)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1103)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1093)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1094)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1037)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.$anonfun$checkType$33(TypeChecker.scala:1362)
[info] [ Debug  ] 	at scala.Option.fold(Option.scala:251)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.$anonfun$checkType$32(TypeChecker.scala:1362)
[info] [ Debug  ] 	at scala.Option.fold(Option.scala:251)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1361)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1276)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.$anonfun$checkType$35(TypeChecker.scala:1375)
[info] [ Debug  ] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
[info] [ Debug  ] 	at scala.collection.Iterator.foreach(Iterator.scala:943)
[info] [ Debug  ] 	at scala.collection.Iterator.foreach$(Iterator.scala:943)
[info] [ Debug  ] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[info] [ Debug  ] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[info] [ Debug  ] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[info] [ Debug  ] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[info] [ Debug  ] 	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
[info] [ Debug  ] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
[info] [ Debug  ] 	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType(TypeChecker.scala:1375)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkType$(TypeChecker.scala:1374)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkType(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkFunctionsAndADTs(TypeChecker.scala:1471)
[info] [ Debug  ] 	at stainless.verification.TypeChecker.checkFunctionsAndADTs$(TypeChecker.scala:1470)
[info] [ Debug  ] 	at stainless.verification.TypeChecker$$anon$2.checkFunctionsAndADTs(TypeChecker.scala:1477)
[info] [ Debug  ] 	at stainless.verification.VerificationRun.$anonfun$execute$3(VerificationComponent.scala:118)
[info] [ Debug  ] 	at scala.util.Try$.apply(Try.scala:213)
[info] [ Debug  ] 	at inox.utils.TimerStorage.runAndGetTime(Timer.scala:93)
[info] [ Debug  ] 	at inox.utils.TimerStorage.run(Timer.scala:88)
[info] [ Debug  ] 	at stainless.verification.VerificationRun.execute(VerificationComponent.scala:118)
[info] [ Debug  ] 	at stainless.verification.VerificationRun.execute(VerificationComponent.scala:53)
[info] [ Debug  ] 	at stainless.ComponentRun.apply(Component.scala:101)
[info] [ Debug  ] 	at stainless.ComponentRun.apply$(Component.scala:98)
[info] [ Debug  ] 	at stainless.verification.VerificationRun.apply(VerificationComponent.scala:53)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.$anonfun$processFunctionsSymbols$2(SplitCallBack.scala:200)
[info] [ Debug  ] 	at scala.collection.immutable.List.map(List.scala:293)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.processFunctionsSymbols(SplitCallBack.scala:199)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.$anonfun$processFunctions$6(SplitCallBack.scala:174)
[info] [ Debug  ] 	at stainless.extraction.utils.ConcurrentCache.getOrElseUpdate(ConcurrentCaches.scala:15)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.processFunctions(SplitCallBack.scala:174)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.$anonfun$processSymbols$4(SplitCallBack.scala:159)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.$anonfun$processSymbols$4$adapted(SplitCallBack.scala:158)
[info] [ Debug  ] 	at scala.collection.immutable.HashMap$HashMapKeys.$anonfun$foreach$1(HashMap.scala:150)
[info] [ Debug  ] 	at scala.collection.immutable.HashMap$HashMap1.foreachEntry(HashMap.scala:400)
[info] [ Debug  ] 	at scala.collection.immutable.HashMap$HashTrieMap.foreachEntry(HashMap.scala:732)
[info] [ Debug  ] 	at scala.collection.immutable.HashMap$HashTrieMap.foreachEntry(HashMap.scala:732)
[info] [ Debug  ] 	at scala.collection.immutable.HashMap$HashMapKeys.foreach(HashMap.scala:150)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.processSymbols(SplitCallBack.scala:158)
[info] [ Debug  ] 	at stainless.frontend.SplitCallBack.endExtractions(SplitCallBack.scala:80)
[info] [ Debug  ] 	at stainless.frontend.ThreadedFrontend$$anon$1.run(ThreadedFrontend.scala:33)
[info] [ Debug  ] 	at java.lang.Thread.run(Thread.java:748)
[info] [ Debug  ] 
[error] Nonzero exit code returned from runner: 2
[error] (stainless-scalac / Compile / run) Nonzero exit code returned from runner: 2
```

</details>

<details>
<summary>Some of the `--debug=trees` output</summary>

It looks like `AntiAliasing inlines the path condition but with identifiers that are not in scope.
```
[info] [ Debug  ] Symbols before AntiAliasing
[info] [ Debug  ] 
[info] [ Debug  ] -------------Functions-------------
[info] [ Debug  ] def get_mut[V @mutable](self: MutCell[Tuple2[String, V @mutable]], (key: String) @pure): Option[V @mutable] = if (self.value._0.value == key) {
[info] [ Debug  ]   Some[V @mutable](self.value._1)
[info] [ Debug  ] } else {
[info] [ Debug  ]   None[V @mutable]()
[info] [ Debug  ] }
[info] [ Debug  ] 
[info] [ Debug  ] @pure
[info] [ Debug  ] def main: Unit = {
[info] [ Debug  ]   val target: MutCell[Tuple2[String, Int]] = MutCell[Tuple2[String, Int]](Tuple2[String, Int](MutCell[String]("bar"), MutCell[Int](0)))
[info] [ Debug  ]   get_mut[Int](target, "foo") match {
[info] [ Debug  ]     case Some(v1) =>
[info] [ Debug  ]       v1.value = 123
[info] [ Debug  ]     case _ =>
[info] [ Debug  ]       ()
[info] [ Debug  ]   }
[info] [ Debug  ] }
[info] [ Debug  ] 
[info] [ Debug  ] -------------Classes-------------
[info] [ Debug  ] case class None[T @mutable] extends Option[T @mutable]
[info] [ Debug  ] 
[info] [ Debug  ] @sealed
[info] [ Debug  ] sealed case class Tuple2[T0 @mutable, T1 @mutable](_0: MutCell[T0 @mutable], _1: MutCell[T1 @mutable])
[info] [ Debug  ] 
[info] [ Debug  ] case class Some[T @mutable](_0: MutCell[T @mutable]) extends Option[T @mutable]
[info] [ Debug  ] 
[info] [ Debug  ] case class MutCell[T @mutable]((value: T @mutable) @var)
[info] [ Debug  ] 
[info] [ Debug  ] @abstract
[info] [ Debug  ] @sealed
[info] [ Debug  ] sealed abstract class Option[T @mutable]
[info] [ Debug  ] 
[info] [ Debug  ] 
[info] [ Debug  ] 
[info] [ Debug  ] 
[info] [ Debug  ] Symbols after AntiAliasing
[info] [ Debug  ] 
[info] [ Debug  ] -------------Functions-------------
[info] [ Debug  ] def get_mut[V @mutable](self: MutCell[Tuple2[String, V @mutable]], (key: String) @pure): Option[V @mutable] = if (self.value._0.value == key) {
[info] [ Debug  ]   Some[V @mutable](self.value._1)
[info] [ Debug  ] } else {
[info] [ Debug  ]   None[V @mutable]()
[info] [ Debug  ] }
[info] [ Debug  ] 
[info] [ Debug  ] @pure
[info] [ Debug  ] def main: Unit = {
[info] [ Debug  ]   var target: MutCell[Tuple2[String, Int]] = MutCell[Tuple2[String, Int]](Tuple2[String, Int](MutCell[String]("bar"), MutCell[Int](0)))
[info] [ Debug  ]   (get_mut[Int](target, "foo") match {
[info] [ Debug  ]     case Some(v1) =>
[info] [ Debug  ]       target = @unchecked (if (target.value._0.value == key) {
[info] [ Debug  ]         MutCell[Tuple2[String, Int]](Tuple2[String, Int](@unchecked @unchecked target.asInstanceOf[MutCell[Tuple2[String, Int]]].value.asInstanceOf[Tuple2[String, Int]]._0, MutCell[Int](123)))
[info] [ Debug  ]       } else {
[info] [ Debug  ]         target
[info] [ Debug  ]       }).asInstanceOf[MutCell[Tuple2[String, Int]]]
[info] [ Debug  ]       ()
[info] [ Debug  ]     case _ =>
[info] [ Debug  ]       ()
[info] [ Debug  ]   })
[info] [ Debug  ] }
[info] [ Debug  ] 
[info] [ Debug  ] -------------Classes-------------
[info] [ Debug  ] case class None[T @mutable] extends Option[T @mutable]
[info] [ Debug  ] 
[info] [ Debug  ] @sealed
[info] [ Debug  ] sealed case class Tuple2[T0 @mutable, T1 @mutable](_0: MutCell[T0 @mutable], _1: MutCell[T1 @mutable])
[info] [ Debug  ] 
[info] [ Debug  ] case class Some[T @mutable](_0: MutCell[T @mutable]) extends Option[T @mutable]
[info] [ Debug  ] 
[info] [ Debug  ] case class MutCell[T @mutable]((value: T @mutable) @var)
[info] [ Debug  ] 
[info] [ Debug  ] @abstract
[info] [ Debug  ] @sealed
[info] [ Debug  ] sealed abstract class Option[T @mutable]
```

</details>

### PR DESCRIPTION
**Pull request by jad-hamza**
Fix #1111

I didn't handle the `LetVar` and `Block` cases, but this seems enough to fix #1111, and a similar issue I was having (see file `WrappedArray.scala`, because we were missing let bindings in (array and map) accessors too.

### CODE REVIEW

**Comment by jad-hamza** [comment]
Merging for now because of the projects that depend on it.

File: core/src/main/scala/stainless/extraction/imperative/AntiAliasing.scala [hunk 1]
```diff
@@ -603,6 +603,9 @@ trait AntiAliasing
         case Target(receiver, None, path) =>
           rec(receiver, path.toSeq)
 
+        case Target(receiver, Some(condition), path) if effects(condition).nonEmpty =>
+          throw FatalError(s"Effects are not allowed in condition of effects: ${condition.asString}")
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/AntiAliasing.scala:607]: Shouldn't this rather be a `MalformedStainlessCode` exception?

File: core/src/main/scala/stainless/extraction/imperative/EffectsAnalyzer.scala [hunk 2]
```diff
@@ -161,30 +161,37 @@ trait EffectsAnalyzer extends oo.CachingPhase {
...
   }
 
   case object UnknownArrayAccessor extends Accessor {
     def asString(implicit ctx: inox.Context) = s"UnknownArrayAccessor"
+    def bind(x: ValDef, e: Expr): Accessor = this
   }
 
   case class MutableMapAccessor(index: Expr) extends Accessor {
     def asString(implicit ctx: inox.Context) = s"MutableMapAccessor(${index.asString})"
+    def bind(x: ValDef, e: Expr): Accessor = ArrayAccessor(bindNonValue(x, e, index))
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/EffectsAnalyzer.scala:189]: Shouldn't this be `MutableMapAccessor` ?

**Review by samarion — COMMENTED** [review]
Sorry for the late review, but I guess it's better late than never.

It seems to me that we have some more general issues with conditional effects. What would happen if we try to pass something like `if (cond) a else b` to a function invocation which has an effect on that argument? Don't we just basically drop the `cond`? That seems wrong to me.

File: core/src/main/scala/stainless/extraction/imperative/EffectsAnalyzer.scala (hunk 2 above)
Comment (jad-hamza) [reply core/src/main/scala/stainless/extraction/imperative/EffectsAnalyzer.scala:189]: Oops, thanks!

File: core/src/main/scala/stainless/extraction/imperative/AntiAliasing.scala (hunk 1 above)
Comment (jad-hamza) [reply core/src/main/scala/stainless/extraction/imperative/AntiAliasing.scala:607]: Yes, thanks!

**Pull request merged**